op_crates/crypto/lib.rs

```diff
@@ -43,6 +81,381 @@ pub fn op_crypto_get_random_values(
+) -> Result<Value, AnyError> {
+  assert_eq!(zero_copy.len(), 1);
+
+  let state = state.borrow();
+  let args: WebCryptoSignArg = serde_json::from_value(args)?;
+  let data = &*zero_copy[0];
+  let algorithm = args.algorithm;
+
+  let signature = match algorithm {
+    Algorithm::RsassaPkcs1v15 => {
+      let resource = state
+        .resource_table
+        .get::<CryptoKeyResource<RSAPrivateKey>>(args.rid)
+        .ok_or_else(bad_resource_id)?;
+
+      let private_key = &resource.key;
+
+      let padding = match resource
+        .hash
+        .ok_or_else(|| WebCryptoError::MissingArgument("hash".to_string()))?
+      {
+        WebCryptoHash::Sha1 => PaddingScheme::PKCS1v15Sign {
+          hash: Some(rsa::hash::Hash::SHA1),
+        },
+        WebCryptoHash::Sha256 => PaddingScheme::PKCS1v15Sign {
+          hash: Some(rsa::hash::Hash::SHA2_256),
+        },
+        WebCryptoHash::Sha384 => PaddingScheme::PKCS1v15Sign {
+          hash: Some(rsa::hash::Hash::SHA2_384),
+        },
+        WebCryptoHash::Sha512 => PaddingScheme::PKCS1v15Sign {
+          hash: Some(rsa::hash::Hash::SHA2_512),
+        },
+      };
+
+      // Sign data based on computed padding and return buffer
+      private_key.sign(padding, &data)?
+    }
+    Algorithm::RsaPss => {
+      let resource = state
+        .resource_table
+        .get::<CryptoKeyResource<RSAPrivateKey>>(args.rid)
+        .ok_or_else(bad_resource_id)?;
+
+      let private_key = &resource.key;
+
+      let rng = OsRng;
+      let salt_len = args.salt_length.ok_or_else(|| {
+        WebCryptoError::MissingArgument("saltLength".to_string())
+      })? as usize;
+
+      let padding = match resource
+        .hash
+        .ok_or_else(|| WebCryptoError::MissingArgument("hash".to_string()))?
+      {
+        WebCryptoHash::Sha1 => {
+          PaddingScheme::new_pss_with_salt::<Sha1, _>(rng, salt_len)
+        }
+        WebCryptoHash::Sha256 => {
+          PaddingScheme::new_pss_with_salt::<Sha256, _>(rng, salt_len)
+        }
+        WebCryptoHash::Sha384 => {
+          PaddingScheme::new_pss_with_salt::<Sha384, _>(rng, salt_len)
+        }
+        WebCryptoHash::Sha512 => {
+          PaddingScheme::new_pss_with_salt::<Sha512, _>(rng, salt_len)
+        }
+      };
+
+      // Sign data based on computed padding and return buffer
+      private_key.sign(padding, &data)?
+    }
+    Algorithm::Ecdsa => {
+      let resource = state
+        .resource_table
+        .get::<CryptoKeyResource<EcdsaKeyPair>>(args.rid)
+        .ok_or_else(bad_resource_id)?;
+      let key_pair = &resource.key;
+
+      // We only support P256-SHA256 & P384-SHA384. These are recommended signature pairs.
+      // https://briansmith.org/rustdoc/ring/signature/index.html#statics
+      if let Some(hash) = args.hash {
+        match hash {
+          WebCryptoHash::Sha256 | WebCryptoHash::Sha384 => (),
+          _ => return Err(WebCryptoError::UnsupportedHash.into()),
+        }
+      };
```

Comment:
ring expects digest algorithm at key generation, so its not possible to apply new hash at signing. Thus we use SHA256 for P256 and SHA384 for P384 at key generation. (This is the recommended signing algorithms as per the [docs](https://briansmith.org/rustdoc/ring/signature/index.html#statics)).